### PR TITLE
Inspect

### PR DESCRIPTION
**Pull request by tummychow**
Capabilities:

- retrieve SHAs from intent and reality stores for all pods on all nodes in the cluster
- filter to a single pod ID or a single node (or both)
- retrieves health check statuses and includes them in the output
- output is unindented json

Pretty printing can be implemented via `p2-inspect | python -mjson.tool`

### CODE REVIEW

**Comment by tummychow** [comment]
Travis is brain farting on this PR as well; tests finished but it never POSTed back to github (in fact even travis itself thinks its tests are still running). Very strange.

**Comment by tummychow** [comment]
looks like the JSON unmarshaling stuff introduced a new bug, one sec while I investigate

File: bin/p2-inspect/main.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+		log.Fatalf("Could not list intent kvpairs: %s", err)
+	}
+	realities, _, err := client.KV().List(reality.REALITY_TREE, nil)
+	if err != nil {
+		log.Fatalf("Could not list reality kvpairs: %s", err)
+	}
+
+	statusMap := make(map[string]map[string]NodePodStatus)
+
+	for _, kvp := range intents {
```
Comment (petertseng) [inline bin/p2-inspect/main.go]: looks like a bit of repeated code when iterating over intents and reality. Worth combining?

Comment (tummychow) [inline bin/p2-inspect/main.go]: yeah I'm going to try to refactor this piece of code once I confirm it works

File: bin/p2-inspect/main.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+var (
+	consulUrl = kingpin.Flag("consul", "The hostname and port of a consul agent in the p2 cluster. Defaults to 0.0.0.0:8500.").String()
+	nodeName  = kingpin.Flag("node", "The node to inspect. By default, all nodes are shown.").String()
+	podId     = kingpin.Flag("pod", "The pod manifest ID to inspect. By default, all pods are shown.").String()
+)
+
+type NodePodStatus struct {
+	NodeName           string `json:"-"`
+	PodId              string `json:"-"`
+	IntentManifestSHA  string
```
Comment (anthonybishopric) [inline bin/p2-inspect/main.go]: I'd suggest converting these fields to `snake_case` in the json marshal. That's consistent with the format of fields in the pod manifest.

File: bin/p2-inspect/main.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+			NodeName:          thisNodeName,
+			PodId:             thisPodId,
+			IntentManifestSHA: thisSHA,
+		}
+	}
+
+	for _, kvp := range realities {
+		splitPoint := strings.LastIndex(kvp.Key, "/")
+		thisNodeName := kvp.Key[len(reality.REALITY_TREE)+1 : splitPoint]
+		thisPodId := kvp.Key[splitPoint+1:]
```
Comment (anthonybishopric) [inline bin/p2-inspect/main.go]: Why not just use the results of `strings.Split`?

**Comment by tummychow** [comment]
finally got to the bottom of the mysterious JSON-encoded YAML bug. Will address the refactors mentioned and push new commits to this PR tomorrow.

**Comment by tummychow** [comment]
Ready for review. More of those hoist launchable test failures, too.

File: bin/p2-inspect/README.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+```bash
+$ p2-inspect | python -mjson.tool
+```
+
+```json
+{
+    "isup": {
+        "aws1.example.com": {
+            "health_check": {
+                "output": "[/data/pods/isup/isup/installs/isup_vsjlzlxvnkizuxmkutqmkqhwukyryztuxhusnkpm/bin/launch]\n[PATH=/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/bin TERM=linux RUNLEVEL=3 PREVLEVEL=N UPSTART_EVENTS=runlevel UPSTART_JOB=runit UPSTART_INSTANCE= CONFIG_PATH=/data/pods/isup/config/isup_717cc0d58df240e2668c865cdc063d715446e6db.yaml]\n",
```
Comment (anthonybishopric) [inline bin/p2-inspect/README.md]: Is your healthcheck for `isup` just echoing `$@`?

File: bin/p2-inspect/README.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            "intent_manifest_sha": "9b9c7cb38b9a68564d6d582c05298cd3dab02e9d",
+            "reality_manifest_sha": "9b9c7cb38b9a68564d6d582c05298cd3dab02e9d"
+        },
+        "aws2.example.com": {
+            "intent_manifest_sha": "26f0f145bf9cc37fa2c337109aa64d7390d8cb4c",
+            "reality_manifest_sha": "26f0f145bf9cc37fa2c337109aa64d7390d8cb4c"
+        },
+        "aws3.example.com": {
+            "intent_manifest_sha": "362c4cc5d28d9dd63e175d5b8889dedbc40483fc",
+            "reality_manifest_sha": "362c4cc5d28d9dd63e175d5b8889dedbc40483fc"
```
Comment (anthonybishopric) [inline bin/p2-inspect/README.md]: Why are the SHAs different on all three nodes? Just wondering

File: bin/p2-inspect/main.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+	}
+
+	for _, kvp := range realities {
+		if addKVPToMap(kvp, REALITY_SOURCE, *filterNodeName, *filterPodId, statusMap) != nil {
+			log.Fatal(err)
+		}
+	}
+
+	for podId := range statusMap {
+		checks, _, err := client.Health().Checks(podId, nil)
```
Comment (anthonybishopric) [inline bin/p2-inspect/main.go]: I'd like to see some of this make its way to the `health` package at some point, at least the Consul part. Part of what this command does is already implemented in the replicator, although definitely with the goal of being informational.

**Comment by anthonybishopric** [comment]
Looking forward to seeing this in action.

File: bin/p2-inspect/README.md (hunk 4 above)
Comment (tummychow) [inline bin/p2-inspect/README.md]: Yeah it echoes `$@` and `env` in the HTTP response. I mainly wanted to confirm (when I was implementing the health checks last week) that the response was getting captured by consul. Seeing it in the output of `p2-inspect` might be useful for debugging.

File: bin/p2-inspect/README.md (hunk 5 above)
Comment (tummychow) [inline bin/p2-inspect/README.md]: They were produced by the `bootstrap` script, which runs `p2-bin2pod` on each node independently.

File: bin/p2-inspect/main.go (hunk 6 above)
Comment (tummychow) [inline bin/p2-inspect/main.go]: Yeah, we are not making much use of the `health` package right now, I think. We can definitely expand on the functionality there.

File: bin/p2-inspect/README.md (hunk 4 above)
Comment (anthonybishopric) [inline bin/p2-inspect/README.md]: Yep totally. That's neat.

File: bin/p2-inspect/README.md (hunk 5 above)
Comment (anthonybishopric) [inline bin/p2-inspect/README.md]: For clarity for new users, just make these the same SHA.

Comment (tummychow) [inline bin/p2-inspect/README.md]: Done

**Comment by tummychow** [comment]
Made one last refactor to health check code, no further action. This is ready for review

**Comment by anthonybishopric** [comment]
:+1:

**Pull request merged**